google/cloud/internal/oauth2_credentials.h

```diff
@@ -88,6 +88,18 @@ class Credentials {
   /// `Options`, is used. Otherwise the default retry policy is used.
   virtual StatusOr<std::string> universe_domain(
       google::cloud::Options const& options) const;
+
+  /**
+   * Return the project (if any) associated with the credentials.
```

Comment:
Is the "if any" part distinct from the error part (like unimplemented)?

Reply:
Good question, let me think on this... While I do that, here are some initial thoughts:

In the cases where a project id never exists (e.g. user credentials), I think returning `StatusCode::kUnimplemented` is fine.  No amount of asking again will return you a project id.

In the cases where a project should exist (e.g. GCE), but getting it failed for some reason, I think returning `StatusCode::kUnavailable` or whatever MDS error we get is the "Right Thing"[tm].

I guess there is the case where a project should exist (e.g. service account credentials), but for whatever reason was missing, then something like `StatusCode::kNotFound` is also fine.

We could express the first and last cases with more subtlety using `StatusOr<absl::optional<std::string>>`.

I think  the action performed by the caller is going to be the same: a missing project id is an error if you need a project id to do something.

Reply:
> We could express the first and last cases with more subtlety using `StatusOr<absl::optional<std::string>>`.
> 
> I think the action performed by the caller is going to be the same: a missing project id is an error if you need a project id to do something.

OK, SGTM.

Then perhaps I might remove the "(if any)", and explicitly state that an error is returned in the "missing" case.

Reply:
Agree. Done.

Reply:
LGTM.